Batch reads on ``/`` fail with coordinator NPE instead of a clear unsupported error

### Search before asking

- [x] I searched in the [issues](https://github.com/apache/fluss/issues) and found nothing similar.

### Fluss version

main (development)

### Please describe the bug 🐞

Batch reads on `$binlog` and `$changelog` virtual tables do not fail fast. Instead, the job is submitted and later fails in the source coordinator with a `NullPointerException`.

This is especially confusing because the failure happens far away from SQL validation and does not explain that virtual log tables are currently stream-oriented.

Reproduction:

1. Create a primary key table with data lake enabled.
2. Run a batch query such as:

```sql
SELECT * FROM `my_table$binlog`;
```

The same problem should also apply to `$changelog`.

Actual behavior:

- The batch job starts.
- `FlinkSourceEnumerator.startInBatchMode()` goes into the hybrid lake split path.
- For virtual tables, the source is created without a `lakeSource`, so `LakeSplitGenerator.generateHybridLakeFlussSplits()` dereferences `lakeSource` and throws `NullPointerException`.

Observed stack trace excerpt:

```text
Caused by: org.apache.flink.util.FlinkRuntimeException: Failed to generate hybrid lake fluss splits
    at org.apache.fluss.flink.source.enumerator.FlinkSourceEnumerator.generateHybridLakeFlussSplits(FlinkSourceEnumerator.java:...)
Caused by: java.lang.NullPointerException
    at org.apache.fluss.flink.lake.LakeSplitGenerator.generateHybridLakeFlussSplits(LakeSplitGenerator.java:118)
```

Relevant code paths:

- `BinlogFlinkTableSource` and `ChangelogFlinkTableSource` create `FlinkSource` as log-only virtual tables.
- `FlinkSourceEnumerator.startInBatchMode()` still routes batch execution to hybrid lake split generation when the base table is lake-enabled.
- `LakeSplitGenerator.generateHybridLakeFlussSplits()` assumes `lakeSource` is non-null.

Expected behavior:

- If batch reads on `$binlog`/`$changelog` are not supported, reject them early with a clear `UnsupportedOperationException` mentioning that these virtual tables only support streaming mode.
- If batch reads are intended to be supported, batch mode should use bounded log-only splits for virtual tables instead of entering the lake hybrid path.

Impact:

- User-visible failure is an internal NPE instead of an actionable error.
- The error is delayed until coordinator split generation, which makes diagnosis unnecessarily difficult.

### Solution

At minimum, add fail-fast validation for `$binlog` and `$changelog` in batch mode so the planner/source creation throws a clear unsupported-mode exception.

A more complete fix would be to make batch virtual-table reads use bounded log-only splits and avoid `LakeSplitGenerator` entirely when `lakeSource == null`.

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/catalog/FlinkTableFactory.java
+++ b/fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/catalog/FlinkTableFactory.java
@@ -322,6 +322,15 @@ public class FlinkTableFactory implements DynamicTableSourceFactory, DynamicTabl
         FlinkConnectorOptionsUtils.validateTableSourceOptions(tableOptions);
     }
 
+    private static void validateVirtualLogTableRuntimeMode(
+            String virtualTableSuffix, boolean isStreamingMode) {
+        if (!isStreamingMode) {
+            throw new UnsupportedOperationException(
+                    String.format(
+                            "%s virtual tables only support streaming mode.", virtualTableSuffix));
+        }
+    }
+
     /** Creates a ChangelogFlinkTableSource for $changelog virtual tables. */
     private DynamicTableSource createChangelogTableSource(
             Context context, ObjectIdentifier tableIdentifier, String tableName) {
@@ -333,6 +342,7 @@ public class FlinkTableFactory implements DynamicTableSourceFactory, DynamicTabl
         boolean isStreamingMode =
                 context.getConfiguration().get(ExecutionOptions.RUNTIME_MODE)
                         == RuntimeExecutionMode.STREAMING;
+        validateVirtualLogTableRuntimeMode(FlinkCatalog.CHANGELOG_TABLE_SUFFIX, isStreamingMode);
 
         // tableOutputType includes metadata columns: [_change_type, _log_offset, _commit_timestamp,
         // data_cols...]
@@ -395,6 +405,7 @@ public class FlinkTableFactory implements DynamicTableSourceFactory, DynamicTabl
         boolean isStreamingMode =
                 context.getConfiguration().get(ExecutionOptions.RUNTIME_MODE)
                         == RuntimeExecutionMode.STREAMING;
+        validateVirtualLogTableRuntimeMode(FlinkCatalog.BINLOG_TABLE_SUFFIX, isStreamingMode);
 
         // tableOutputType: [_change_type, _log_offset, _commit_timestamp, before ROW<...>, after
         // ROW<...>]

--- a/fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/catalog/FlinkTableFactoryTest.java
+++ b/fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/catalog/FlinkTableFactoryTest.java
@@ -245,11 +245,9 @@ abstract class FlinkTableFactoryTest {
     private ResolvedSchema createBinlogSchema() {
         return new ResolvedSchema(
                 Arrays.asList(
-                        Column.physical(
-                                "_change_type", DataTypes.STRING().notNull()),
+                        Column.physical("_change_type", DataTypes.STRING().notNull()),
                         Column.physical("_log_offset", DataTypes.BIGINT().notNull()),
-                        Column.physical(
-                                "_commit_timestamp", DataTypes.TIMESTAMP_LTZ(3).notNull()),
+                        Column.physical("_commit_timestamp", DataTypes.TIMESTAMP_LTZ(3).notNull()),
                         Column.physical(
                                 "before",
                                 DataTypes.ROW(
@@ -289,11 +287,7 @@ abstract class FlinkTableFactoryTest {
             Map<String, String> options,
             Map<String, String> enrichmentOptions) {
         return createTableSource(
-                OBJECT_IDENTIFIER,
-                schema,
-                options,
-                enrichmentOptions,
-                new Configuration());
+                OBJECT_IDENTIFIER, schema, options, enrichmentOptions, new Configuration());
     }
 
     private static DynamicTableSource createTableSource(

--- a/fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/catalog/FlinkTableFactoryTest.java
+++ b/fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/catalog/FlinkTableFactoryTest.java
@@ -23,7 +23,9 @@ import org.apache.fluss.flink.source.FlinkTableSource;
 import org.apache.fluss.flink.source.lookup.FlinkAsyncLookupFunction;
 import org.apache.fluss.flink.source.lookup.FlinkLookupFunction;
 
+import org.apache.flink.api.common.RuntimeExecutionMode;
 import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.ExecutionOptions;
 import org.apache.flink.table.api.DataTypes;
 import org.apache.flink.table.api.Schema;
 import org.apache.flink.table.api.ValidationException;
@@ -70,6 +72,12 @@ abstract class FlinkTableFactoryTest {
 
     public static final ObjectIdentifier OBJECT_IDENTIFIER =
             ObjectIdentifier.of("default", "default", "t1");
+
+    public static final ObjectIdentifier CHANGELOG_TABLE_IDENTIFIER =
+            ObjectIdentifier.of("default", "default", "t1" + FlinkCatalog.CHANGELOG_TABLE_SUFFIX);
+
+    public static final ObjectIdentifier BINLOG_TABLE_IDENTIFIER =
+            ObjectIdentifier.of("default", "default", "t1" + FlinkCatalog.BINLOG_TABLE_SUFFIX);
 
     @Test
     void testTableSourceOptions() {
@@ -180,6 +188,36 @@ abstract class FlinkTableFactoryTest {
     }
 
     @Test
+    void testVirtualLogTableSourceDoesNotSupportBatchMode() {
+        ResolvedSchema schema = createBasicSchema();
+        Map<String, String> properties = getBasicOptions();
+        Configuration configuration = new Configuration();
+        configuration.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.BATCH);
+
+        assertThatThrownBy(
+                        () ->
+                                createTableSource(
+                                        CHANGELOG_TABLE_IDENTIFIER,
+                                        schema,
+                                        properties,
+                                        Collections.emptyMap(),
+                                        configuration))
+                .isInstanceOf(UnsupportedOperationException.class)
+                .hasMessageContaining("$changelog virtual tables only support streaming mode.");
+
+        assertThatThrownBy(
+                        () ->
+                                createTableSource(
+                                        BINLOG_TABLE_IDENTIFIER,
+                                        createBinlogSchema(),
+                                        properties,
+                                        Collections.emptyMap(),
+                                        configuration))
+                .isInstanceOf(UnsupportedOperationException.class)
+                .hasMessageContaining("$binlog virtual tables only support streaming mode.");
+    }
+
+    @Test
     void testSink() {
         ResolvedSchema schema = createBasicSchema();
         Map<String, String> properties = getBasicOptionsWithBucketKey();
@@ -204,6 +242,30 @@ abstract class FlinkTableFactoryTest {
                 UniqueConstraint.primaryKey("PK_first_third", Arrays.asList("first", "third")));
     }
 
+    private ResolvedSchema createBinlogSchema() {
+        return new ResolvedSchema(
+                Arrays.asList(
+                        Column.physical(
+                                "_change_type", DataTypes.STRING().notNull()),
+                        Column.physical("_log_offset", DataTypes.BIGINT().notNull()),
+                        Column.physical(
+                                "_commit_timestamp", DataTypes.TIMESTAMP_LTZ(3).notNull()),
+                        Column.physical(
+                                "before",
+                                DataTypes.ROW(
+                                        DataTypes.FIELD("first", DataTypes.STRING().notNull()),
+                                        DataTypes.FIELD("second", DataTypes.INT()),
+                                        DataTypes.FIELD("third", DataTypes.STRING().notNull()))),
+                        Column.physical(
+                                "after",
+                                DataTypes.ROW(
+                                        DataTypes.FIELD("first", DataTypes.STRING().notNull()),
+                                        DataTypes.FIELD("second", DataTypes.INT()),
+                                        DataTypes.FIELD("third", DataTypes.STRING().notNull())))),
+                Collections.emptyList(),
+                null);
+    }
+
     private static Map<String, String> getBasicOptions() {
         Map<String, String> options = new HashMap<>();
         options.put("connector", "fluss");
@@ -226,10 +288,24 @@ abstract class FlinkTableFactoryTest {
             ResolvedSchema schema,
             Map<String, String> options,
             Map<String, String> enrichmentOptions) {
+        return createTableSource(
+                OBJECT_IDENTIFIER,
+                schema,
+                options,
+                enrichmentOptions,
+                new Configuration());
+    }
+
+    private static DynamicTableSource createTableSource(
+            ObjectIdentifier objectIdentifier,
+            ResolvedSchema schema,
+            Map<String, String> options,
+            Map<String, String> enrichmentOptions,
+            Configuration configuration) {
         FlinkTableFactory tableFactory = createFlinkTableFactory();
         FactoryUtil.DefaultDynamicTableContext context =
                 new FactoryUtil.DefaultDynamicTableContext(
-                        OBJECT_IDENTIFIER,
+                        objectIdentifier,
                         new ResolvedCatalogTable(
                                 CatalogTable.of(
                                         Schema.newBuilder().fromResolvedSchema(schema).build(),
@@ -240,7 +316,7 @@ abstract class FlinkTableFactoryTest {
                                         options),
                                 schema),
                         enrichmentOptions,
-                        new Configuration(),
+                        configuration,
                         Thread.currentThread().getContextClassLoader(),
                         false);
         return tableFactory.createDynamicTableSource(context);

--- a/fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/source/BinlogVirtualTableITCase.java
+++ b/fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/source/BinlogVirtualTableITCase.java
@@ -131,6 +131,15 @@ abstract class BinlogVirtualTableITCase {
 
     // init table environment from savepointPath
     private StreamTableEnvironment initTableEnvironment(@Nullable String savepointPath) {
+        return initTableEnvironment(savepointPath, EnvironmentSettings.inStreamingMode());
+    }
+
+    private StreamTableEnvironment initBatchTableEnvironment() {
+        return initTableEnvironment(null, EnvironmentSettings.inBatchMode());
+    }
+
+    private StreamTableEnvironment initTableEnvironment(
+            @Nullable String savepointPath, EnvironmentSettings environmentSettings) {
         org.apache.flink.configuration.Configuration conf =
                 new org.apache.flink.configuration.Configuration();
         if (savepointPath != null) {
@@ -140,8 +149,7 @@ abstract class BinlogVirtualTableITCase {
                 StreamExecutionEnvironment.getExecutionEnvironment(conf);
         execEnv.setParallelism(1);
         execEnv.enableCheckpointing(1000);
-        StreamTableEnvironment tEnv =
-                StreamTableEnvironment.create(execEnv, EnvironmentSettings.inStreamingMode());
+        StreamTableEnvironment tEnv = StreamTableEnvironment.create(execEnv, environmentSettings);
         String bootstrapServers = String.join(",", clientConf.get(ConfigOptions.BOOTSTRAP_SERVERS));
         // crate catalog using sql
         tEnv.executeSql(
@@ -225,6 +233,22 @@ abstract class BinlogVirtualTableITCase {
         // $binlog should fail for log tables
         assertThatThrownBy(() -> tEnv.executeSql("DESCRIBE log_table$binlog").collect())
                 .hasMessageContaining("only supported for primary key tables");
+    }
+
+    @Test
+    public void testBatchReadBinlogTableFailsFast() throws Exception {
+        tEnv.executeSql(
+                "CREATE TABLE batch_binlog_test ("
+                        + "  id INT NOT NULL,"
+                        + "  name STRING,"
+                        + "  PRIMARY KEY (id) NOT ENFORCED"
+                        + ") WITH ('bucket.num' = '1')");
+
+        tEnv = initBatchTableEnvironment();
+
+        assertThatThrownBy(() -> tEnv.explainSql("SELECT * FROM batch_binlog_test$binlog"))
+                .hasRootCauseInstanceOf(UnsupportedOperationException.class)
+                .hasRootCauseMessage("$binlog virtual tables only support streaming mode.");
     }
 
     @Test

--- a/fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/source/ChangelogVirtualTableITCase.java
+++ b/fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/source/ChangelogVirtualTableITCase.java
@@ -64,6 +64,7 @@ import static org.apache.fluss.flink.utils.FlinkTestBase.writeRows;
 import static org.apache.fluss.testutils.DataTestUtils.row;
 import static org.apache.fluss.testutils.common.CommonTestUtils.waitUntil;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 /** Integration test for $changelog virtual table functionality. */
 abstract class ChangelogVirtualTableITCase {
@@ -128,6 +129,15 @@ abstract class ChangelogVirtualTableITCase {
 
     // init table environment from savepointPath
     private StreamTableEnvironment initTableEnvironment(@Nullable String savepointPath) {
+        return initTableEnvironment(savepointPath, EnvironmentSettings.inStreamingMode());
+    }
+
+    private StreamTableEnvironment initBatchTableEnvironment() {
+        return initTableEnvironment(null, EnvironmentSettings.inBatchMode());
+    }
+
+    private StreamTableEnvironment initTableEnvironment(
+            @Nullable String savepointPath, EnvironmentSettings environmentSettings) {
         org.apache.flink.configuration.Configuration conf =
                 new org.apache.flink.configuration.Configuration();
         if (savepointPath != null) {
@@ -137,8 +147,7 @@ abstract class ChangelogVirtualTableITCase {
                 StreamExecutionEnvironment.getExecutionEnvironment(conf);
         execEnv.setParallelism(1);
         execEnv.enableCheckpointing(1000);
-        StreamTableEnvironment tEnv =
-                StreamTableEnvironment.create(execEnv, EnvironmentSettings.inStreamingMode());
+        StreamTableEnvironment tEnv = StreamTableEnvironment.create(execEnv, environmentSettings);
         String bootstrapServers = String.join(",", clientConf.get(ConfigOptions.BOOTSTRAP_SERVERS));
         // crate catalog using sql
         tEnv.executeSql(
@@ -237,6 +246,22 @@ abstract class ChangelogVirtualTableITCase {
                                 + "  `tags` ARRAY<VARCHAR(2147483647)>\n"
                                 // with options contains random properties, skip checking
                                 + ")");
+    }
+
+    @Test
+    public void testBatchReadChangelogTableFailsFast() throws Exception {
+        tEnv.executeSql(
+                "CREATE TABLE batch_changelog_test ("
+                        + "  id INT NOT NULL,"
+                        + "  name STRING,"
+                        + "  PRIMARY KEY (id) NOT ENFORCED"
+                        + ") WITH ('bucket.num' = '1')");
+
+        tEnv = initBatchTableEnvironment();
+
+        assertThatThrownBy(() -> tEnv.explainSql("SELECT * FROM batch_changelog_test$changelog"))
+                .hasRootCauseInstanceOf(UnsupportedOperationException.class)
+                .hasRootCauseMessage("$changelog virtual tables only support streaming mode.");
     }
 
     @Test

